test/test_ops.py

```diff
@@ -279,106 +280,198 @@ def _ref_test_helper(
-            torch_result = op.torch_opinfo(sample.input, *sample.args, **sample.kwargs)
-
-            for a, b in zip(tree_flatten(ref_result)[0], tree_flatten(torch_result)[0]):
-                if isinstance(a, torch.Tensor) or isinstance(b, torch.Tensor):
-                    prims.utils.compare_tensor_meta(a, b)
-                    if getattr(op, 'validate_view_consistency', True) and not skip_view_consistency:
-                        msg = (f"The torch implementation {'returns' if b._is_view() else 'does not return'} "
-                               f"a view, while the reference {'does' if a._is_view() else 'does not'}")
-                        self.assertEqual(a._is_view(), b._is_view(), msg)
-
-            # Computes the dtype the more precise computatino would occur in
-            precise_dtype = torch.bool
-            if prims.utils.is_integer_dtype(dtype):
-                # Note: bool and integer dtypes do not have more
-                # precise dtypes -- they simply must be close
-                precise_dtype = dtype
-            if prims.utils.is_float_dtype(dtype):
-                precise_dtype = torch.double
-            if prims.utils.is_complex_dtype(dtype):
-                precise_dtype = torch.cdouble
-
-            # Checks if the results are close
-            try:
-                self.assertEqual(
-                    ref_result,
-                    torch_result,
-                    exact_stride=False,
-                    exact_device=True,
-                    exact_layout=True,
-                    exact_is_coalesced=True,
-                )
-            except AssertionError as e:
-                # Raises the error if the precise dtype comparison wouldn't be
-                # different
-                if dtype is precise_dtype:
-                    raise e
 
-                ex = e
+            # Check op behavior and API compatibility inside and outside ctx.
+            test_inputs = make_test_inputs(op)
+
+            # No matching test input functions found, exit right away instead of
+            # iterating through the rest of sample inputs.
+            if not test_inputs:
+                self.skipTest("Skipped! No matching test inputs found")
+
+            for test_input_ctx, test_input_no_ctx in test_inputs:
+                with ctx() as c:
+                    ref_result = test_input_ctx(sample.input, *sample.args, **sample.kwargs)
+                    if hasattr(c, "mock") and c.mock:
+                        # This tests API compatibility by checking whether we
+                        # can call a ref with sample inputs by going through the
+                        # torch to ref dispatch. The mocked ref returns a
+                        # sentinel value, so move to the next sample instead of
+                        # doing any consistency checks against the no_ctx
+                        # result.
+                        self.assertEqual(ref_result, MOCK_REF_RESULT)
```

Comment:
Rather than the comment above, could you add a short message that the dev would get if this assert fails summarising what's the error and what should they do?

Reply:
Done. I also noticed this assert is not always reached because dispatch may fail earlier, so I added exception handling there, which prints a user-friendly error message on failure.